tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,380 @@
+#include "tachyon/crypto/commitments/fri/two_adic_fri_prover.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_verifier.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree.h"
+#include "tachyon/crypto/commitments/mixed_matrix_commitment_scheme.h"
+#include "tachyon/math/finite_fields/extension_field_traits_forward.h"
+#include "tachyon/math/geometry/dimensions.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/math/matrix/matrix_utils.h"
+#include "tachyon/math/polynomials/univariate/evaluations_utils.h"
+#include "tachyon/zk/air/plonky3/base/two_adic_multiplicative_coset.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename ExtField, typename InputMMCS, typename ChallengeMMCS,
+          typename Challenger>
+class TwoAdicFriPCS {
+ public:
+  using F = typename math::ExtensionFieldTraits<ExtField>::BaseField;
+  using Coset = zk::air::plonky3::TwoAdicMultiplicativeCoset<F>;
+  using Commitment = typename InputMMCS::Commitment;
+  using ProverData = typename InputMMCS::ProverData;
+  using Proof = typename InputMMCS::Proof;
+  using InputProof = std::vector<BatchOpening<F, InputMMCS>>;
+  using TwoAdicFriProof =
+      crypto::TwoAdicFriProof<ExtField, ChallengeMMCS, F, InputProof>;
+
+  using Points = std::vector<std::vector<ExtField>>;
+
+  using OpenedValuesForMat = std::vector<std::vector<ExtField>>;
+  using OpenedValuesForRound = std::vector<OpenedValuesForMat>;
+  using OpenedValues = std::vector<OpenedValuesForRound>;
+
+  TwoAdicFriPCS() = default;
+  TwoAdicFriPCS(InputMMCS& mmcs, TwoAdicFriConfig<ChallengeMMCS>& fri)
+      : mmcs_(mmcs), fri_(fri) {}
+
+  Coset GetNaturalDomainForDegree(size_t degree) {
+    size_t log_n = base::bits::StrictLog2(degree);
+    return Coset(log_n, F::One());
+  }
+
+  [[nodiscard]] bool Commit(std::vector<Coset>& cosets,
+                            std::vector<math::RowMajorMatrix<F>>& matrices,
+                            Commitment* commitment, ProverData* prover_data) {
+    std::vector<math::RowMajorMatrix<F>> ldes(cosets.size());
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < cosets.size(); ++i) {
+      Coset& coset = cosets[i];
+      math::RowMajorMatrix<F>& mat = matrices[i];
+      CHECK_EQ(coset.domain()->size(), static_cast<size_t>(mat.rows()));
+      coset.domain()->CosetLDEBatch(
+          mat, fri_.log_blowup,
+          unwrap(F::FromMontgomery(F::Config::kSubgroupGenerator) /
+                 coset.domain()->offset()));
+      ReverseMatrixIndexBits(mat);
+      ldes[i] = mat;
+    }
+    return mmcs_.Commit(std::move(ldes), commitment, prover_data);
+  }
+
+  Eigen::Block<math::RowMajorMatrix<F>> GetEvaluationsOnDomain(
+      size_t idx, const ProverData& prover_data, Coset& coset) {
+    CHECK_EQ(coset.domain()->offset(),
+             F::FromMontgomery(F::Config::kSubgroupGenerator));
+    math::RowMajorMatrix<F>& lde = mmcs_.GetMatrices(prover_data)[idx];
```

Comment:
How it's compiled..? `GetMatrices()` returns a const reference.

Reply:
That is a very good point 🤔 I'm not sure...

Reply:
This is because `GetEvaluationsOnDomain()` not used any where. Please add unittest for this to instantiate this template method or remove this!

Reply:
I'll remove it~